### PR TITLE
Shutdown plugins during daemon shutdown.

### PR DESCRIPTION
**Pull request by anusha-ragunathan**
Signed-off-by: Anusha Ragunathan anusha@docker.com

### CODE REVIEW

**Comment by cpuguy83** [comment]
But we shouldn't do this if live-restore is enabled, right?

**Comment by cpuguy83** [comment]
daemon/daemon.go:687: undefined: plugin.GetManager

**Comment by anusha-ragunathan** [comment]
@cpuguy83 : The live-restore case bails out much early in the function.

File: plugin/manager_linux.go [hunk 1]
```diff
@@ -128,3 +128,21 @@ func (pm *Manager) disable(p *plugin) error {
 	pm.save()
 	return nil
 }
+
+// Shutdown stops all plugins and called during daemon shutdown.
+func (pm *Manager) Shutdown() error {
```
Comment (thaJeztah) [inline plugin/manager_linux.go]: We never return an error, correct? Should we fail if a plugin fails to stop?

File: plugin/manager_linux.go [hunk 2]
```diff
@@ -128,3 +128,21 @@ func (pm *Manager) disable(p *plugin) error {
...
+			if err := p.restartManager.Cancel(); err != nil {
+				logrus.Error(err)
+			}
+		}
+		if pm.containerdClient != nil {
+			if err := pm.containerdClient.Signal(p.P.ID, int(syscall.SIGKILL)); err != nil {
+				logrus.Error(err)
+			}
+		}
+		os.RemoveAll(p.runtimeSourcePath)
```
Comment (thaJeztah) [inline plugin/manager_linux.go]: Should we check / log errors here?

**Comment by thaJeztah** [comment]
@anusha-ragunathan is this for 1.12?

File: plugin/manager_linux.go (hunk 1 above)
Comment (anusha-ragunathan) [inline plugin/manager_linux.go]: We return the error now. But its purely for logging purposes. The daemon shutdown should not fail, if any of the sub components fail to stop. It should continue its best-effort shutdown, as it does today.

**Comment by anusha-ragunathan** [comment]
@thaJeztah : Yes, we should take this for 1.12

**Comment by LK4D4** [comment]
@anusha-ragunathan Is it possible to write test?

**Comment by anusha-ragunathan** [comment]
@LK4D4 : Done!

**Comment by LK4D4** [comment]
@anusha-ragunathan great, thanks!

**Comment by LK4D4** [comment]
I think experimental failure is unrelated.
LGTM

**Comment by cpuguy83** [comment]
```
daemon/daemon_experimental.go:15: plugin.GetManager().Shutdown undefined (type *plugin.Manager has no field or method Shutdown)
```

**Comment by anusha-ragunathan** [comment]
@cpuguy83 : Fixed the experimental windows cross compile issue. PTAL.
Experimental and Windows native are failing across the board.

File: plugin/manager_linux.go [hunk 3]
```diff
@@ -128,3 +128,23 @@ func (pm *Manager) disable(p *plugin) error {
...
+// Shutdown stops all plugins and called during daemon shutdown.
+func (pm *Manager) Shutdown() error {
+	for _, p := range pm.plugins {
+		if p.restartManager != nil {
+			if err := p.restartManager.Cancel(); err != nil {
+				return err
+			}
+		}
+		if pm.containerdClient != nil {
+			if err := pm.containerdClient.Signal(p.P.ID, int(syscall.SIGKILL)); err != nil {
```
Comment (cpuguy83) [inline plugin/manager_linux.go]: We should give the plugin a chance to shutdown gracefully with `SIGTERM`

Comment (cpuguy83) [inline plugin/manager_linux.go]: Or rather `STOPSIGNAL` if that was brought over from the image format.

**Comment by anusha-ragunathan** [comment]
@cpuguy83 : Graceful shutdown added. PTAL.

**Comment by cpuguy83** [comment]
LGTM

**Pull request merged**